### ISSUE TITLE
Full validation of chain candidates with respect to the chain they extend

### ISSUE THREAD

**Issue by adinapoli-mndc** [oscoin/oscoin#420]
See https://github.com/oscoin/oscoin/pull/407#discussion_r267220722

In a nutshell, #402 left out two things during chain selection:

- We didn't prune chains which were selected but that failed full-validation;
- We didn't validate chains with respect to the chain they extend. By that I mean that we validate a chain only "locally", but not "globally", i.e. see if it validates once we "attach" it to the main chain, and we validate, say, `<last_k_blocks_of_main_chain + new_fork>`.

**Comment by cloudhead** [comment]
On your second point, what you're saying for eg. is given two chains:
```
genesis -> a -> b                          and                             c -> d -> e
```
We validate `c -> d -> e`, but not `c` with respect to `b` for example.

**Comment by adinapoli-mndc** [comment]
Right, exactly -- I am concerned that if we don't validate the _whole_ chain, as in

```
genesis -> a -> b -> c -> d -> e
```

How we can be sure that the new main chain is correct?

> I wonder what the minimum is that we need to do? Is it <mutable_blocks_of_main_chain + fork> or is it just <mutable_blocks_of_main_chain + fork[0]>

Yeah, good question. I suppose this will be the point where it would have been useful to have a mathematician / logician between us (do we have one?) so that we could have tried formalising some basic properties, prove them (via Coq or more simply via pen & paper), derive some simple lemma and finally write quickcheck properties out of that 😉 

My intuition (but I am just a humble engineer 😁 ) is that our invariant should be that `immutable_blocks_of_main_chain` is correct, so is `mutable_blocks_of_main_chain`. The line of reasoning is that, if `mutable_blocks_of_main_chain` is invalid that is an absurd, as our algorithm shouldn't have picked those blocks in the first place. Then, by induction, we can build up more interesting properties on it. Probably if we validate `mutable_blocks_of_main_chain + fork`, this should suffice, me thinks.

/EndOfHandwaving
